module.go

```diff
@@ -5,15 +5,40 @@ type Scanner interface {
 	// Init runs once for this module at library init time
 	Init(flags ScanFlags) error
 
-	// InitPerSender runs once per Goroutine. A single Goroutine will scan some non-deterministics
+	// InitPerSender runs once per Goroutine. A single Goroutine will scan some non-deterministic
 	// subset of the input scan targets
 	InitPerSender(senderID int) error
 
 	// Returns the name passed at init
 	GetName() string
 
 	// Scan connects to a host. The result should be JSON-serializable
-	Scan(t ScanTarget) (interface{}, error)
+	Scan(t ScanTarget) (ScanStatus, interface{}, error)
+}
+
+// ScanStatus is the enum value that states how the scan ended.
+type ScanStatus string
+
+// TODO: Confirm to standard string const format (names, capitalization, hyphens/underscores, etc)
+// TODO: Enumerate further status types
+const (
+	SCAN_SUCCESS            = "success"            // The protocol in question was positively identified and the scan encountered no errors
+	SCAN_CONNECTION_REFUSED = "connection-refused" // TCP connection was actively rejected
+	SCAN_CONNECTION_TIMEOUT = "connection-timeout" // No response to TCP connection request
+	SCAN_CONNECTION_CLOSED  = "connection-closed"  // The TCP connection was unexpectedly closed
+	SCAN_IO_TIMEOUT         = "io-timeout"         // Timed out waiting on data
+	SCAN_PROTOCOL_ERROR     = "protocol-error"     // Received data incompatible with the target protocol
+	SCAN_APPLICATION_ERROR  = "application-error"  // The application reported an error
+	SCAN_UNKNOWN_ERROR      = "unknown-error"      // Catch-all for unrecognized errors
+)
+
+// ScanResponse is the result of a scan on a single host
+type ScanResponse struct {
+	// Status is required for all responses. Other fields are optional.
+	Status    ScanStatus  `json:"status"`
+	Result    interface{} `json:"result,omitempty"`
```

Comment:
Shouldn't that be possible with a custom un-marshaller? i.e. `MyClass.Unmarshal(input map[string]interface{}) ret *MyClass`? It's tedious, as hell to write, but it would only be problematic if you had `interface{}` members further nested inside the class -- but then that exists for that class regardless of whether `ScanResponse` has `interface{} Result`s or not.

Without some type metadata, you obviously won't get be able to unmarshal mystery data without knowing the type in advance (the same problem cropped up with zschema -- it's currently cheating by assuming that the protocol name is the key into the data map, so validating scans on the same service on multiple ports of a single machine is not possible).

Reply:
> It's tedious, as hell to write, but it would only be problematic if you had interface{} members further nested inside the class.

This ends up being tedious enough that you just never use Go to unmarshal things. I guess in the meantime, if you know your own schema, you can just create a matching struct without `interface{}` and unmarshal it that way. The way we've gotten around this previously (at least to some extent), is to have the `interface{}` field instead be a union of all possible options into one struct, and then just enforce manually in code that only one member is ever non-nil.